internal/coreinternal/golden/normalize_timestamps.go

```diff
@@ -0,0 +1,151 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package golden // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/golden"
+
+import (
+	"sort"
+	"time"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil"
+)
+
+func normalizeTimestamps(metrics pmetric.Metrics) {
+	rms := metrics.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		for j := 0; j < rms.At(i).ScopeMetrics().Len(); j++ {
+			for k := 0; k < rms.At(i).ScopeMetrics().At(j).Metrics().Len(); k++ {
+				m := rms.At(i).ScopeMetrics().At(j).Metrics().At(k)
+				switch m.Type() {
+				case pmetric.MetricTypeGauge:
+					normalizeDataPointSlice(dataPointSlice[pmetric.NumberDataPoint](m.Gauge().DataPoints()))
+				case pmetric.MetricTypeSum:
+					normalizeDataPointSlice(dataPointSlice[pmetric.NumberDataPoint](m.Sum().DataPoints()))
+				case pmetric.MetricTypeHistogram:
+					normalizeDataPointSlice(dataPointSlice[pmetric.HistogramDataPoint](m.Histogram().DataPoints()))
+				case pmetric.MetricTypeExponentialHistogram:
+					normalizeDataPointSlice(dataPointSlice[pmetric.ExponentialHistogramDataPoint](m.ExponentialHistogram().DataPoints()))
+				case pmetric.MetricTypeSummary:
+					normalizeDataPointSlice(dataPointSlice[pmetric.SummaryDataPoint](m.Summary().DataPoints()))
+				}
+			}
+		}
+	}
+}
+
+type timeStampInfo struct {
+	timestamp pcommon.Timestamp
+	position  int
+}
```

Comment:
I don't think we need this struct, or any of the position tracking.

All we need is:
- Sort a `[]pcommon.Timestamp` that contains all `Timestamp` and `StartTimestamp`, regardless of uniqueness
- Create `normalizedTs` the same as you are doing
- Iterate over all DataPoints, and use `normalizeTs` to determine the appropriate value to set each timestamp
```go
for _, dp := range dps {
    dp.SetTimestamp(normalizeTs[dp.Timestamp()])
    dp.SetStartTimestamp(normalizeTs[dp.StartTimestamp()])
}
```